### PR TITLE
Asg 3355 msaad unarchive

### PR DESCRIPTION
**Pull request by bosvos**
Add unarchive mechanism to MSAAD

You'll see 3 commits here.
- The first (oldest) is the feature implementation only with very few tests
- The second is an interface and implementation refactor of the MSAAD type, so that we can write proper tests without hitting MS Graph. There may have been *slight overkill* here and the MSAAD interface iteself may not have been necessary. The MSAADProvider interface *is* though. Your call.
- The final commit is just a cleanup, moving methods around as appropriate

### CODE REVIEW

**Review by FritzOnFire — CHANGES_REQUESTED** [review]
Had some concerns about how the child is depended on the parent, but that seems pretty deeply ingrained in this library, so I am not going to ask to fix that now.

I think my big changes was around the tests, and how we should rather make use of what we have as much as possible, instead of the very handcrafted code that was made in `msaad_test.go`

File: msaad.go [hunk 1]
```diff
@@ -18,97 +18,197 @@ package authaus
...
+// MSAADProviderI
+//
+// Interface to abstract the fetching of roles and users, allowing mocking
+// of returns.
+// Initialize must be called on all implementing structs.
+type MSAADProviderI interface {
+	fetchJSON(fetchURL string, jsonRoot interface{}) error
+	doLoggedHTTP(request *http.Request) (*http.Response, []byte, error)
+	doHTTP(request *http.Request) (*http.Response, error)
+	getBearerToken() (token string, expiresAt time.Time, err error)
```
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad.go:73]: The issue is that we need a way to separate e.g. getAADUsers from the calling function (SynchronizeUsers) which is under test...to be able to test SynchronizeUsers, we need a concrete instance of MSAAD, but we don't want to call its own getAADUsers for this, because we want to supply our own...any ideas?

File: msaad.go [hunk 2]
```diff
@@ -18,97 +18,197 @@ package authaus
...
+//
+// If _both_ GivenName and Surname are populated, then we use those.
+//
+// If not, we attempt to split the DisplayName into a name and surname.
 func (u *msaadUserJSON) nameAndSurname() (string, string) {
 	if u.GivenName != "" && u.Surname != "" {
 		return u.GivenName, u.Surname
 	}
 	return splitDisplayName(u.DisplayName)
 }
```
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad.go:91]: Yes was thinking about that, just figured that it also sometimes takes _other_ unrelated parameters (like AuthUser), so could not decide where to put it.

File: msaad.go [hunk 3]
```diff
@@ -292,34 +349,54 @@ func (crg *cachedRoleGroups) idToGroupName(i GroupIDU32) string {
...
+	}
 	m.parent = parent
-	m.tokenExpiresAt = time.Now().Add(-time.Hour)
+	if log == nil {
+		return fmt.Errorf("MSAAD logger is nil")
+	}
+
+	m.log = log
+	msaadProvider := MSAADProvider{}
+	err := msaadProvider.Initialize(m, m.log)
```
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad.go:392]: You think we should refrain from referencing "parent" in the child...so we just need to push the references in as fields rather.

File: msaad.go [hunk 4]
```diff
@@ -292,34 +349,54 @@ func (crg *cachedRoleGroups) idToGroupName(i GroupIDU32) string {
...
+	if err != nil {
+		return fmt.Errorf("could not initialise MSAAD provider")
+	}
+	msaadProvider.tokenExpiresAt = time.Now().Add(-time.Hour)
```
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad.go:396]: This is pre-existing code from initialisation of MSAAD, and since it is effectively a global on the MSAAD itself, it was probably just a safeguard to not have a default value that is actually valid. Possibly came out during testin years ago and was just set to be in the past by default:

```
func (m *MSAAD) Initialize(parent *Central) {
	m.parent = parent
	m.tokenExpiresAt = time.Now().Add(-time.Hour)
}
```

File: msaad.go [hunk 5]
```diff
@@ -333,20 +410,20 @@ func (m *MSAAD) SynchronizeUsers() error {
 
 	// Log errors about missing internal group names (this is a config mistake)
 	// We do this check once during sync, to avoid emitting this error during the sync of every user.
-	for aadRole, internalGroupName := range m.Config.RoleToGroup {
+	for aadRole, internalGroupName := range m.Config().RoleToGroup {
```
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad.go:413]: I agree...I'll clean that up...I think I did realise that towards the end of the exercise and forgot to clean that up.

File: auth.go [hunk 6]
```diff
@@ -769,13 +787,18 @@ func (x *Central) MergeTick() {
 		MergeLDAP(x)
 	}
 	if x.msaadSyncMergeEnabled {
+		// Seriously consider removing the count check here
+		if x.syncMergeCount%1 == 0 {
```
[the remark this answers is not in the document]
Comment (bosvos) [reply auth.go:791]: Apologies, must have set it to 1 to get more visibility.

File: auth.go [hunk 7]
```diff
@@ -769,13 +787,18 @@ func (x *Central) MergeTick() {
...
+		}
 		if err := x.MSAAD.SynchronizeUsers(); err != nil {
 			x.Log.Warnf("MSAAD synchronization failed: %v", err)
 		}
 	}
 
 	timeComplete := time.Now()
-	if x.syncMergeCount%60 == 0 {
+	// Seriously consider removing the count check here
+	if x.syncMergeCount%1 == 0 {
```
[the remark this answers is not in the document]
Comment (bosvos) [reply auth.go:801]: Same as prior.

File: msaad.go [hunk 8]
```diff
@@ -375,10 +452,25 @@ func (m *MSAAD) SynchronizeUsers() error {
...
+			// 1 - the user was local/LDAP and needs to be moved to MSAAD
+			//		are we really going to just trust that it is the same user?
+			// 2 - the user's got a NEW account in MSAAD and was re-invited,
+			//		in which case we also should not assume anything about the user...
+			//		especially since MSAAD would not be assuming anything about the user.
+			// 3 - there is a third possibility, moving the guest tenant (us),
+			//		in which case the email and home-tenant UPN would be the same.
+			//		Is that enough to assume that the user is the same?
+			// So, i would vote for removing this line. ALL users should start
+			// from scratch when we move over to MSAAD.???
```
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad.go:473]: Keeping it for now. With the comment.

File: msaad.go [hunk 9]
```diff
@@ -18,97 +18,197 @@ package authaus
...
+// Interface to abstract the fetching of roles and users, allowing mocking
+// of returns.
+// Initialize must be called on all implementing structs.
+type MSAADProviderI interface {
+	fetchJSON(fetchURL string, jsonRoot interface{}) error
+	doLoggedHTTP(request *http.Request) (*http.Response, []byte, error)
+	doHTTP(request *http.Request) (*http.Response, error)
+	getBearerToken() (token string, expiresAt time.Time, err error)
+	getAADUsers() ([]*msaadUser, error)
+	getUserAssignments(user *msaadUser, i int) (errGlobal error, quit bool)
```
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad.go:75]: Same as before, except getAADUsers and getUserAssignments are used to supply users. I'll change them to non-private.

File: msaad.go [hunk 10]
```diff
@@ -785,172 +840,66 @@ func (m *MSAAD) populateAADRoles(users []*msaadUser) error {
...
+	nThreads := nItems / 10
+	if nThreads < 1 {
+		nThreads = 1
 	}
-	if err != nil {
-		m.parent.Log.Errorf("MSAAD failed to %v %v (err): %v", request.Method, request.URL.String(), err)
-	} else if response.StatusCode != 200 {
-		m.parent.Log.Errorf("MSAAD failed to %v %v (response): %v %v", request.Method, request.URL.String(), response.Status, string(responseBody))
+	if nThreads > 8 {
+		nThreads = 8
```
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad.go:886]: I'm not sure what the original intent was with this limit...but I don't see it necessary to create two threads for any available CPU. Unless you feel strongly about it...but this does translate to actual HTTP calls and perhaps we should not be flooding the Graph API?

File: msaad.go [hunk 11]
```diff
@@ -785,172 +840,66 @@ func (m *MSAAD) populateAADRoles(users []*msaadUser) error {
...
-	token = tokenJSON.AccessToken
-	// can time.Duration(string) not PANIC?
-	expiresAt = time.Now().Add(time.Duration(tokenJSON.ExpiresIn) * time.Second)
-
-	if m.Config.Verbose {
-		m.parent.Log.Infof("MSAAD bearer token refreshed successfully: '%v'. ExpiresAt: %v", token[:4], expiresAt)
-	}
-
-	return
+func removeFromGroupList(list []GroupIDU32, i int) []GroupIDU32 {
```
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad.go:900]: This was pre-existing code, from here still: 269f5eb6 Ben Harper <rogojin@gmail.com> on 8/14/2020 at 11:00 AM
Did not focus much on the older stuff unless I had a need for a new parameter etc.

File: msaad.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad.go:91]: Moved it.

File: msaad_entraid.go [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+		return
+	}
+
+	if tokenJSON.TokenType != "Bearer" {
+		err = fmt.Errorf("Unexpected MSAAD token type '%v' (expected 'Bearer')", tokenJSON.TokenType)
+		return
+	}
+
+	token = tokenJSON.AccessToken
+	// can time.Duration(string) not PANIC?
```
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad_entraid.go:272]: Yeah sorry, missed that.

File: msaad_test.go [hunk 13]
```diff
@@ -0,0 +1,879 @@
...
+how the user's "Username" is treated.
+*/
+
+type TestMSAAD struct {
+	parent        *Central
+	log           *log.Logger
+	msaad         MSAAD
+	dummyConfig   ConfigMSAAD
+	dummyProvider MSAADProviderI
+}
```
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad_test.go:30]: Yes. I guess I was in too deep. I've removed and move around some initialisation code.

File: msaad_test.go [hunk 14]
```diff
@@ -0,0 +1,879 @@
...
+		assert.Equal(t, io.out, io.target, "Failed for %v", io.in)
+	}
+}
+
+func Test_SynchronizeUsers(t *testing.T) {
+	c := getCentralMSAAD(t)
+	e := c.MSAAD.SynchronizeUsers()
+	assert.Equal(t, 1, len(c.userStore.(*sanitizingUserStore).backend.(*testUserStore).usersUpdated))
+	assert.Equal(t, 2, len(c.userStore.(*sanitizingUserStore).backend.(*testUserStore).usersArchived))
+	assert.Equal(t, 1, len(c.userStore.(*sanitizingUserStore).backend.(*testUserStore).usersUnarchived))
```
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad_test.go:870]: As you mentioned earlier, some of these implemented functions are not nicely tied together. Will rework the userstore to be more consistent.

File: msaad_test.go [hunk 15]
```diff
@@ -0,0 +1,879 @@
...
+
+func (t *testRoleGroupDB) UpdateGroup(group *AuthGroup) error {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (t *testRoleGroupDB) Close() {
+	//TODO implement me
+	panic("implement me")
+}
```
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad_test.go:614]: Yes initially I think they were different. Will move over.

File: msaad_test.go [hunk 16]
```diff
@@ -0,0 +1,879 @@
...
+			ID:   5,
+			Name: "enabled",
+			PermList: PermissionList{
+				5, 6,
+			},
+		},
+	}
+	return d
+
+}
```
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad_test.go:521]: Yup.

File: msaad_test.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad_test.go:870]: Just a note. One benefit of using testUserStore is that you can test if a function is entered or not. While you can devise a scheme to _deduct_ if this is the case (compare before and after), you cannot prove it. Normally that's fine (we can test e.g. "Archived" for all users and easily prove that only the correct ones were changed). For "Update", it is more difficult and you would have to compare ALL users in play fully to also cater for the negative case (i.e. update was called but it wasn't supposed to). In both cases it is expensive to have a certain level of confidence. I guess targeted testing is OK, but it does not give you 100% surety.

File: db_sql.go [hunk 17]
```diff
@@ -520,6 +520,72 @@ func (x *sqlUserStoreDB) ArchiveIdentity(userId UserId) error {
...
+	}
+	if err = tx.Commit(); err != nil {
+		return nil, fmt.Errorf("Could not commit transaction: %w", err)
+	}
+	// return the user object
+	aUser, err := getUserFromUserId(x.db, id)
+	if err != nil {
+		return nil, err
+	}
+	return aUser, nil
```
[the remark this answers is not in the document]
Comment (bosvos) [reply db_sql.go:546]: Sorry haven't done this yet. Will do in next PR.

File: msaad.go [hunk 18]
```diff
@@ -389,53 +481,80 @@ func (m *MSAAD) SynchronizeUsers() error {
...
+						continue
+					} else {
+						internalUserID = newUserID
+						if m.parent.Auditor != nil {
+							contextData := userInfoToAuditTrailJSON(user, "")
+							m.parent.Auditor.AuditUserAction(user.Username, "User Profile: "+user.Username, contextData, AuditActionCreated)
+						}
+					}
 				}
 			}
```
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad.go]: Sorry not done yet, will do in next PR.

File: msaad.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (FritzOnFire) [reply msaad.go]: I was thinking more in lines of `Central` has a reference to `MSAADProviderI` and `MSAADProviderI` as a reference to `MSAADInterface`, that way the children don't know about their parents. But after some digging I saw that `Central` is also passed to its children :( So we would need a lot more changes here... Starting with putting `Central` in its own package, and `LDAP` in its own packge, and `MSAAD` is its own package. Then go would automatically shout at us about cyclic dependencies and we would be forced to fix it.

Should probably tackle that in a different PR.

File: msaad.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (FritzOnFire) [reply msaad.go:715]: I feel strongly about creating two threads per CPU. This is the equivalent to `downloadThreads` that we have in some of our services. Very little utilization on our side, mostly just waiting.

I would very much like to flood Graph API. Not in a malicious way. But in a, "the overhead of this call is more than the execution of this call" kinda way (at lest, it should be).

File: msaad_test.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (FritzOnFire) [reply msaad_test.go]: I get where you are coming from... I would just rather use more fine grained unit tests to drill into the edge cases. Which I guess we should do anyway to ensure that we don't have regression on any of the most basic functionality.

File: msaad_entraid.go [hunk 19]
```diff
@@ -0,0 +1,444 @@
...
+			mp.log.Errorf("MSAAD failed to read response body from %v %v: %v", request.Method, request.URL.String(), err)
+			return response, nil, err
+		}
+	}
+	if err != nil {
+		mp.log.Errorf("MSAAD failed to %v %v (err): %v", request.Method, request.URL.String(), err)
+	} else if response.StatusCode != 200 {
+		mp.log.Errorf("MSAAD failed to %v %v (response): %v %v", request.Method, request.URL.String(), response.Status, string(responseBody))
+	}
+	return response, responseBody, err
```
Comment (FritzOnFire) [inline msaad_entraid.go:202]: I don't know why this method is so hard to read... please just change it to this.
```suggestion
	response, err := mp.doHTTP(request)
	if err != nil {
		e := fmt.Errorf("MSAAD failed to %v %v (err): %w", request.Method, request.URL.String(), err)
		mp.log.Error(e.Error())
		return nil, nil, e
	}

	defer response.Body.Close()
	body, err := io.ReadAll(response.Body)
	if err != nil {
		e := fmt.Errorf("MSAAD failed to read response body from %v %v: %w", request.Method, request.URL.String(), err)
		mp.log.Error(e.Error())
		return nil, nil, e
	}

	if response.StatusCode != 200 {
		e := fmt.Errorf("MSAAD failed to %v %v (response): %v %v", request.Method, request.URL.String(), response.Status, string(responseBody))
		mp.log.Error(e.Error())
		return nil, nil, e
	}

	return response, body, nil
```
This also has the benefit of returning much better errors.

File: msaad_entraid.go [hunk 20]
```diff
@@ -0,0 +1,444 @@
...
+	if err != nil {
+		return fmt.Errorf("Error creating Request object for url '%v': %v", fetchURL, err)
+	}
+	response, body, err := mp.doLoggedHTTP(request)
+	if err != nil {
+		return fmt.Errorf("Error fetching '%v' (err): %w", fetchURL, err)
+	}
+	if response.StatusCode != 200 {
+		return fmt.Errorf("Error fetching '%v' (response): %v", fetchURL, response.Status)
+	}
```
Comment (FritzOnFire) [inline msaad_entraid.go:177]: You should not need this check anymore after changing `doLoggedHTTP`

File: msaad_test.go [hunk 21]
```diff
@@ -0,0 +1,559 @@
...
+
+	groupIds := GroupIDU32s{GroupIDU32(0), GroupIDU32(1), GroupIDU32(2)}
+	var p Permit
+	p.Roles = EncodePermit(groupIds)
+	permitDB.SetPermit(1, &p)
+	p.Roles = EncodePermit(groupIds)
+	permitDB.SetPermit(2, &p)
+
+	roleDB = newDummyRoleGroupDB()
+	if err := roleDB.InsertGroup(&AuthGroup{
```
Comment (FritzOnFire) [inline msaad_test.go:62]: Right! Just shove these function calls into this method! Why did I not think about that. No need for more methods, just have it all in one place. Very nice :)

**Review by FritzOnFire — CHANGES_REQUESTED** [review]
Some small changes, aside from the function change. But I think we are pretty much there.

File: msaad.go [hunk 22]
```diff
@@ -470,31 +382,69 @@ func (m *MSAAD) SynchronizeUsers() error {
 	return nil
 }
 
+func (m *MSAAD) CreateOrUnarchiveUser(aadUser *msaadUser, aadEmailClean string) UserId {
```
[the remark this answers is not in the document]
Comment (bosvos) [reply msaad.go:385]: Will have to call bestemail again there, or move the logging down.

[the remark this answers is not in the document]
Comment (bosvos) [reply msaad.go:385]: Perhaps it is just easier to write out the full msaadUser...but it is an "Info" log...so will cause a bit of extra logging.

[the remark this answers is not in the document]
Comment (bosvos) [reply msaad.go:385]: Just added both mail and userprincipalname there. One reason for the logging is to be clear on the identifying details of the msaad user...since there were some misalignment and misidentification in the past.

[the remark this answers is not in the document]
Comment (FritzOnFire) [reply msaad.go]: NOOOOOOOOOOOOOOO. You have a line `user := aadUser.profile.toAuthUser()`. The second line in `toAuthUser()` is literally a call to `bestEmail()` which is stored on `Email` on the `user`... please just use the `Email` on `user`...

**Review by FritzOnFire — APPROVED** [review]
My only remaining request would be to change the `CreateOrUnarchiveUser` function to use  `user.Email` for logging, but I am ok, if we don't make that change. LGTM
